### ISSUE TITLE
Discrepancies in Weight Optimization and Inference

### ISSUE THREAD

**Issue by vlntlbr** [amzn/BayesPE#1]
Hello all,

Thanks for the interesting method and for sharing the code. I have been reviewing the implementation and found a few discrepancies that might affect the performance and calibration of the model:

In `src/ensemble_scaler.py`, the entropy term in the loss function is currently implemented as:
`entropy = - torch.dot(scales.double(), torch.exp(scales).double())`
Since scales represents a probability distribution, `exp(scales) `will always be positive and greater than 1, meaning the term does not represent information-theoretic entropy and will lead to incorrect regularization during weight optimization.

In `src/bpe.py`, the forward method allows for selecting a subset of instructions using n_forward_passes. The code retrieves the top-$k$ weights and passes them directly to the scaler: `chosen_weights = np.sort(self.weights)[-n_forward_passes:]return self.scaler.scale_ensemble(probs, chosen_weights)`.

Because `chosen_weights` is a subset of the original weights, they do not necessarily sum to 1.0. Since the `scale_ensemble` function performs a weighted sum of probabilities, the resulting output distribution will not sum to 1.0 if the weights are not re-normalized. 

Also, the likelihood calculation in `src/ensemble_scaler.py` is implemented as the log of the weighted sum of probabilities:` likelihood = torch.sum(torch.log(torch.matmul(probs_y.double(), scales.double())+SMALL_CONSTANT))`. The BayesPE paper frames weight estimation as a Variational Inference problem, which usually maximizes the Evidence Lower Bound (ELBO). This objective generally involves the expectation of the log-likelihood ($\sum w_i \log p_i$) rather than the log of the expectation.

I believe fixing the entropy term and adding weight normalization in the inference pass would greatly improve the reliability of this implementation.

Best,
Valentina
